cli/src/integrations/neon/auth.ts

```diff
@@ -0,0 +1,102 @@
+import { log } from "@clack/prompts";
+import { createServer } from "node:http";
+import type { AddressInfo } from "node:net";
+import {
+  type TokenResponseBody,
+  OAuth2Client,
+  generateState,
+  generateCodeVerifier,
+} from "oslo/oauth2";
+import open from "open";
+
+const NEON_OAUTH_HOST = "https://oauth2.neon.tech";
+const NEON_CLIENT_ID = "create-honc-app";
+const NEON_SCOPES = [
+  "openid",
+  "offline",
+  "offline_access",
+  "urn:neoncloud:projects:read",
+  "urn:neoncloud:projects:create",
+];
+const AUTH_TIMEOUT_MS = 200000;
+const REDIRECT_URI = (port: number) => `http://127.0.0.1:${port}/callback`;
+
+export async function getNeonAuthToken(): Promise<TokenResponseBody> {
+  const server = createServer();
+  server.listen(0, "127.0.0.1");
+  await new Promise<void>((resolve) => server.once("listening", resolve));
+  const listenPort = (server.address() as AddressInfo).port;
+
+  const client = new OAuth2Client(
+    NEON_CLIENT_ID,
+    `${NEON_OAUTH_HOST}/oauth2/auth`,
+    `${NEON_OAUTH_HOST}/oauth2/token`,
+    { redirectURI: REDIRECT_URI(listenPort) },
+  );
+
+  const state = generateState();
+  const codeVerifier = generateCodeVerifier();
+
+  const authorizationURL = await client.createAuthorizationURL({
+    state,
+    codeVerifier,
+    codeChallengeMethod: "S256",
+    scopes: NEON_SCOPES,
+  });
+
+  return new Promise((resolve, reject) => {
+    const timer = setTimeout(() => {
+      reject(
+        new Error(
+          `Authentication timed out after ${AUTH_TIMEOUT_MS / 1000} seconds`,
+        ),
+      );
+    }, AUTH_TIMEOUT_MS);
+
+    server.on("request", async (req, res) => {
+      if (!req.url?.startsWith("/callback")) {
+        res.writeHead(404);
+        res.end();
+        return;
+      }
+
+      const url = new URL(req.url, REDIRECT_URI(listenPort));
+      const code = url.searchParams.get("code");
+      const returnedState = url.searchParams.get("state");
+
+      if (returnedState !== state) {
+        reject(new Error("State mismatch"));
+        return;
+      }
+
+      if (code) {
+        try {
+          const token = await client.validateAuthorizationCode(code, {
+            codeVerifier,
+          });
+          res.writeHead(200, { "Content-Type": "text/html" });
+          res.end(
+            "<code>Authentication with create-honc-app successful! You can close this window.</code>",
```

Comment:
goose? 🙈

Reply:
OBLIGATORY

Reply:
actually having tested it out, it won't work on plain text responses (punting it for the next update :))